### ISSUE TITLE
training perf for single GPU is not good

### ISSUE THREAD

**Issue by cyy857** [mlfoundations/open_clip#13]
Hi, I was training clip using single GPU. After profiling, I noticed that the perf of CLIP training was not good, as we can see from the picture below. GPU idle time is almost twice of GPU active due to the sem_timedwait as blocked in CPU. Any idea we can solve this unnecessary block? Thanks!
![image](https://user-images.githubusercontent.com/16970791/132194276-c60552ac-9fbd-4316-8216-72d5593e9e68.png)

**Comment by gabrielilharco** [comment]
Hi @cyy857,

Can you provide more details of your setup?
In particular, what kind of dataset and what GPU are you using, and the values of the following params:

- batch size
- workers
- precision
- gpu
- dp
- multigpu

Thanks!

**Comment by cyy857** [comment]
Thanks   @gabrielilharco , I used bs=128, workers=1, precision=amp, single GPU (V100), not using dp/multigpu. 

After increasing batch size to 150 (about 1.5G mem was occupied during training) and workers to 8, for single GPU training ,the GPU utils became very high then.

From my previous understanding, the workers influence dataloader only, right? But it seems related to the sem_timedwait as shown on above picture? Thanks for any insight about this!

**Comment by mitchellnw** [comment]
> After increasing batch size to 150 (about 1.5G mem was occupied during training) and workers to 8, for single GPU training ,the GPU utils became very high then.

This is what I would recommend doing. I think we mostly just experimented with increasing batch-size / workers until GPU utils looked good. Not sure exactly what `sem_timedwait` is, do you mind elaborating?

**Comment by gabrielilharco** [comment]
Can't say for sure, but this sounds like you weren't feeding data to the gpus fast enough, which is why increasing the number of workers helped.

**Comment by rwightman** [comment]
The code has changed quite a bit and I've done testing on 2x - 4x GPU with decent performance on V100, RTX3090, and Titan RTX cards for the ResNet50 models. So please try again if you were still having problems. I typically use 4-6 workers per GPU. 

Also note that the accuracy of results improves significantly with larger batch sizes afforded by using more cards (more negative examples across the global batch). Hopefully we get a change to include gradient caching soon so this can be achieved in lower GPU settings (but with extended training time).
